### ISSUE TITLE
Task 1 & 2: Flood Segmentation + Burn Scar Segmentation on Jetson Orin Nano

### ISSUE THREAD

**Issue by vikashkodati** [godel-space/prithvi-edge-benchmark#1]
## Overview

Benchmark the Prithvi-EO-2.0-300M foundation model on two disaster response tasks using **pre-trained checkpoints** (no training required). Run inference on NVIDIA Jetson Orin Nano and measure accuracy, latency, power consumption, and memory usage across multiple precision formats (FP32, FP16, INT8).

These two tasks were chosen first because **fine-tuned checkpoints already exist on HuggingFace** -- this is pure inference benchmarking, not model training.

---

## Hardware Requirements

- **NVIDIA Jetson Orin Nano 8GB Developer Kit** ($499)
  - JetPack 6.x (latest stable)
  - "Super Mode" enabled (67 TOPS, 25W max)
- **Storage:** 128GB+ microSD or NVMe SSD (datasets + models need ~20GB)
- **Power measurement:** One of:
  - USB power meter on the barrel jack
  - `jtop` Python package (`pip install jetson-stats`) -- reads onboard power sensors
  - `tegrastats` (built into JetPack) -- logs power at configurable interval

---

## Software Environment Setup

```bash
# On Jetson Orin Nano with JetPack 6.x installed

# 1. Create virtual environment
python3 -m venv .venv
source .venv/bin/activate

# 2. Install PyTorch for Jetson (use NVIDIA's pre-built wheel)
# Check https://developer.nvidia.com/embedded/jetson-zoo for latest wheel URL
pip install numpy
pip install torch torchvision  # Use Jetson-specific wheel from NVIDIA

# 3. Install dependencies
pip install terratorch           # IBM's fine-tuning framework for Prithvi
pip install onnx onnxruntime-gpu # ONNX export and runtime
pip install rasterio             # GeoTIFF reading
pip install scikit-learn         # Metrics (confusion matrix, F1)
pip install matplotlib           # Visualization
pip install jetson-stats         # Power monitoring (jtop)
pip install huggingface_hub      # Model/dataset download

# 4. Verify TensorRT is available (comes with JetPack)
python3 -c "import tensorrt; print(tensorrt.__version__)"

# 5. Verify CUDA
python3 -c "import torch; print(torch.cuda.is_available(), torch.cuda.get_device_name(0))"
```

---

### 1.1 Download Checkpoint

```bash
# Pre-trained Prithvi 300M fine-tuned on Sen1Floods11
# HuggingFace: https://huggingface.co/ibm-nasa-geospatial/Prithvi-EO-2.0-300M-TL-Sen1Floods11
from huggingface_hub import snapshot_download
snapshot_download(
    repo_id="ibm-nasa-geospatial/Prithvi-EO-2.0-300M-TL-Sen1Floods11",
    local_dir="./models/flood-seg"
)
```

### 1.2 Download Dataset

```bash
# Sen1Floods11 dataset
# Source: https://github.com/cloudtostreet/Sen1Floods11
# Alternative: https://huggingface.co/datasets/ibm-nasa-geospatial/sen1floods11
git clone https://github.com/cloudtostreet/Sen1Floods11.git ./data/sen1floods11

# Dataset details:
# - 4,831 image chips (512 x 512 pixels)
# - 446 hand-labeled chips + 4,385 weakly-labeled chips
# - Sentinel-1 SAR + Sentinel-2 optical
# - 11 flood events across 6 continents, 14 biomes
# - Coverage: 120,406 km²
# - Labels: 0 = no water, 1 = water/flood, -1 = no data
```

### 1.3 Run Inference -- PyTorch FP32 Baseline

```python
# run_flood_benchmark.py (skeleton -- adapt based on TerraTorch API)
import torch
import time
import numpy as np
from pathlib import Path

# Load model
model = ...  # Load from TerraTorch or HuggingFace checkpoint
model.eval()
model.cuda()

# Load test split images
test_images = [...]  # Load 512x512 GeoTIFF chips from test split

# Warmup (5 iterations, discard)
for img in test_images[:5]:
    with torch.no_grad():
        _ = model(img.cuda())
torch.cuda.synchronize()

# Benchmark
latencies = []
predictions = []
for img, label in test_set:
    torch.cuda.synchronize()
    start = time.perf_counter()
    with torch.no_grad():
        pred = model(img.unsqueeze(0).cuda())
    torch.cuda.synchronize()
    elapsed_ms = (time.perf_counter() - start) * 1000
    latencies.append(elapsed_ms)
    predictions.append(pred.argmax(dim=1).cpu().numpy())

# Report
print(f"Mean latency: {np.mean(latencies):.1f} ms/tile")
print(f"Median latency: {np.median(latencies):.1f} ms/tile")
print(f"P95 latency: {np.percentile(latencies, 95):.1f} ms/tile")
print(f"Throughput: {1000/np.mean(latencies):.2f} tiles/sec")
```

### 1.4 Export to ONNX

```python
# export_onnx.py
import torch

model = ...  # Load checkpoint
model.eval()

dummy_input = torch.randn(1, 6, 512, 512).cuda()  # 6 bands, 512x512
# Note: Check actual input shape from model card -- may be (1, 6, T, 224, 224) 
# where T = number of time steps. Verify from checkpoint config.

torch.onnx.export(
    model,
    dummy_input,
    "models/flood-seg/model.onnx",
    opset_version=17,
    input_names=["input"],
    output_names=["output"],
    dynamic_axes={"input": {0: "batch"}, "output": {0: "batch"}}
)

# Validate ONNX
import onnx
onnx_model = onnx.load("models/flood-seg/model.onnx")
onnx.checker.check_model(onnx_model)
print(f"ONNX model size: {Path('models/flood-seg/model.onnx').stat().st_size / 1e6:.1f} MB")
```

### 1.5 Build TensorRT FP16 Engine

```bash
# On Jetson (TensorRT comes with JetPack)
/usr/src/tensorrt/bin/trtexec \
    --onnx=models/flood-seg/model.onnx \
    --saveEngine=models/flood-seg/model_fp16.engine \
    --fp16 \
    --workspace=4096 \
    --verbose \
    2>&1 | tee logs/trtexec_flood_fp16.log

# Record the build log -- it shows layer-by-layer optimization details
```

### 1.6 Build TensorRT INT8 Engine (if feasible)

```bash
# INT8 requires calibration data (subset of training images)
# Create calibration cache first, then build engine
/usr/src/tensorrt/bin/trtexec \
    --onnx=models/flood-seg/model.onnx \
    --saveEngine=models/flood-seg/model_int8.engine \
    --int8 \
    --calib=calibration_cache.bin \
    --workspace=4096 \
    --verbose

# If INT8 calibration is too complex, skip this step and note "INT8 not attempted"
```

### 1.7 Run TensorRT Inference + Measure

```python
# Same benchmark loop as 1.3 but using TensorRT engine
import tensorrt as trt
import pycuda.driver as cuda
import pycuda.autoinit

# Load engine, allocate buffers, run inference, measure latency
# Record same metrics as FP32 baseline
```

### 1.8 Power Measurement

```bash
# Option A: jtop (recommended)
# In a separate terminal, start logging:
jtop --csv power_log.csv

# Then run the benchmark in another terminal
# After benchmark completes, stop jtop
# Parse power_log.csv for:
#   - Peak GPU power (W)
#   - Peak total board power (W)
#   - Average GPU power during inference (W)
#   - Average total board power during inference (W)

# Option B: tegrastats
tegrastats --interval 100 --logfile tegrastats_log.txt &
# Run benchmark
# Kill tegrastats
# Parse log for VDD_GPU_SOC and VDD_CPU_CV power rails
```

---

### 2.1 Download Checkpoint

```bash
# Pre-trained Prithvi 300M fine-tuned on HLS burn scars
# HuggingFace: https://huggingface.co/ibm-nasa-geospatial/Prithvi-EO-2.0-300M-BurnScars
from huggingface_hub import snapshot_download
snapshot_download(
    repo_id="ibm-nasa-geospatial/Prithvi-EO-2.0-300M-BurnScars",
    local_dir="./models/burn-scar"
)
```

### 2.2 Download Dataset

```bash
# HLS Burn Scars dataset
# HuggingFace: https://huggingface.co/datasets/ibm-nasa-geospatial/hls_burn_scars
from huggingface_hub import snapshot_download
snapshot_download(
    repo_id="ibm-nasa-geospatial/hls_burn_scars",
    repo_type="dataset",
    local_dir="./data/hls_burn_scars"
)

# Dataset details:
# - 804 image chips (512 x 512 pixels)
# - 6 HLS bands (Blue, Green, Red, NIR, SWIR1, SWIR2)
# - 30m spatial resolution
# - Labels: binary (burn scar / no burn scar)
# - Source: NASA Harmonized Landsat Sentinel-2 (HLS)
# - Ground truth from MTBS (Monitoring Trends in Burn Severity)
```

### 2.3 Run Same Pipeline as Task 1

The pipeline is identical:
1. PyTorch FP32 baseline inference on test split
2. ONNX export (opset 17)
3. TensorRT FP16 engine build
4. TensorRT INT8 engine build (if feasible)
5. Benchmark each configuration
6. Power measurement

The only differences:
- Input: 6-band HLS imagery (not SAR)
- Checkpoint: BurnScars model (not Sen1Floods11)
- Labels: binary segmentation (burn scar vs. background)

---

### Accuracy Metrics

| Metric | Definition | How to Compute |
|--------|-----------|----------------|
| **mIoU** | Mean Intersection over Union across all classes | `sklearn.metrics.jaccard_score(y_true, y_pred, average='macro')` |
| **Per-class IoU** | IoU for each class separately (e.g., flood class IoU, background IoU) | `jaccard_score(y_true, y_pred, average=None)` |
| **F1 score** | Harmonic mean of precision and recall | `sklearn.metrics.f1_score(y_true, y_pred, average='macro')` |
| **Per-class F1** | F1 for each class separately | `f1_score(y_true, y_pred, average=None)` |
| **Overall Accuracy** | Percentage of correctly classified pixels | `sklearn.metrics.accuracy_score(y_true.flatten(), y_pred.flatten())` |
| **Precision** | True positives / (true positives + false positives) | For the target class (flood / burn scar) |
| **Recall** | True positives / (true positives + false negatives) | For the target class (flood / burn scar) |

**Important:** Ignore pixels with label = -1 (no data) when computing metrics. Mask them out before metric computation.

### Performance Metrics

| Metric | Definition | How to Measure |
|--------|-----------|----------------|
| **Mean latency (ms/tile)** | Average inference time per 512x512 tile | `time.perf_counter()` with `torch.cuda.synchronize()` before start/stop |
| **Median latency (ms/tile)** | Median inference time (more robust than mean) | Sort latencies, take middle value |
| **P95 latency (ms/tile)** | 95th percentile latency (worst-case planning) | `np.percentile(latencies, 95)` |
| **Throughput (tiles/sec)** | How many tiles processed per second | `1000 / mean_latency_ms` |
| **Peak GPU power (W)** | Maximum GPU power draw during inference | From `jtop` or `tegrastats` log |
| **Average GPU power (W)** | Mean GPU power during inference batch | From `jtop` or `tegrastats` log |
| **Peak board power (W)** | Maximum total system power | From `jtop` or `tegrastats` log |
| **Average board power (W)** | Mean total system power during inference | From `jtop` or `tegrastats` log |
| **Model file size (MB)** | Size of model file on disk | For each format: PyTorch (.pt), ONNX (.onnx), TRT (.engine) |
| **Peak GPU memory (MB)** | Maximum GPU RAM used during inference | `torch.cuda.max_memory_allocated()` or from `jtop` |
| **Energy per tile (mJ)** | Energy consumed per inference | `average_power_W * mean_latency_s * 1000` |

### Accuracy Degradation Metrics

| Metric | Definition |
|--------|-----------|
| **FP16 vs FP32 accuracy delta** | mIoU(FP16) - mIoU(FP32). Should be < 0.01 (< 1%) |
| **INT8 vs FP32 accuracy delta** | mIoU(INT8) - mIoU(FP32). Acceptable if < 0.03 (< 3%) |
| **FP16 speedup** | latency(FP32) / latency(FP16). Expect 1.5-3x |
| **INT8 speedup** | latency(FP32) / latency(INT8). Expect 2-4x |

---

#### 1. Results Table (Markdown)

```markdown
## Task: [Flood Segmentation / Burn Scar Segmentation]
- **Dataset:** [name] ([N] test chips, [resolution])
- **Checkpoint:** [HuggingFace URL]
- **Hardware:** Jetson Orin Nano 8GB, JetPack [X.Y], CUDA [X.Y], TensorRT [X.Y]
- **Date:** [YYYY-MM-DD]

| Config | mIoU | Target IoU | F1 | OA | Latency (ms) | P95 (ms) | Throughput | GPU Pwr (W) | Board Pwr (W) | Model (MB) | GPU RAM (MB) | Energy/tile (mJ) |
|--------|------|------------|----|----|-------------|----------|------------|-------------|---------------|------------|-------------|------------------|
| PyTorch FP32 | | | | | | | | | | | | |
| ONNX FP32 | | | | | | | | | | | | |
| TRT FP16 | | | | | | | | | | | | |
| TRT INT8 | | | | | | | | | | | | |

### Accuracy Degradation
| Comparison | mIoU Delta | F1 Delta | Speedup |
|------------|-----------|----------|---------|
| FP16 vs FP32 | | | |
| INT8 vs FP32 | | | |
```

#### 2. Visual Outputs (PNG screenshots)

For **3 representative test images** per task:
- Side-by-side comparison: `[Input Image] | [Ground Truth Mask] | [FP32 Prediction] | [TRT FP16 Prediction]`
- Save as `results/flood_sample_01.png`, `results/flood_sample_02.png`, etc.
- Choose images that show: (a) a good prediction, (b) a challenging case, (c) a failure case

#### 3. Raw Data Files

- `results/flood_latencies_fp32.json` -- list of all per-tile latency measurements
- `results/flood_latencies_fp16.json` -- same for TensorRT FP16
- `results/flood_power_log.csv` -- power measurement log during inference
- `results/flood_predictions/` -- directory of all predicted masks (optional, for later analysis)

#### 4. Environment Info

```bash
# Capture and save
python3 -c "
import torch, platform, subprocess
print(f'Python: {platform.python_version()}')
print(f'PyTorch: {torch.__version__}')
print(f'CUDA: {torch.version.cuda}')
print(f'cuDNN: {torch.backends.cudnn.version()}')
try:
    import tensorrt; print(f'TensorRT: {tensorrt.__version__}')
except: print('TensorRT: not available')
try:
    import onnxruntime; print(f'ONNX Runtime: {onnxruntime.__version__}')
except: print('ONNX Runtime: not available')
" > results/environment.txt

# Also capture Jetson info
sudo jetson_clocks --show >> results/environment.txt
cat /etc/nv_tegra_release >> results/environment.txt
```

---

### Task 1: Flood Segmentation
| Metric | Target | Cloud SOTA Reference |
|--------|--------|---------------------|
| mIoU (TRT FP16) | > 0.65 | 0.73 (bi-temporal SAR, cloud GPU) |
| Flood class IoU | > 0.60 | ~0.67 (Attentive U-Net) |
| Latency (TRT FP16) | < 3,000 ms per 512x512 tile | N/A (cloud has no power constraint) |
| Power | < 15W average board power | N/A |

### Task 2: Burn Scar Segmentation
| Metric | Target | Cloud SOTA Reference |
|--------|--------|---------------------|
| mIoU (TRT FP16) | > 0.70 | 0.73 (Prithvi 100M, cloud GPU) |
| Burn scar class IoU | > 0.68 | ~0.73 |
| Latency (TRT FP16) | < 3,000 ms per 512x512 tile | N/A |
| Power | < 15W average board power | N/A |

---

## Known Gotchas / Notes

1. **Input shape:** Prithvi EO 2.0 expects multi-temporal inputs. The checkpoint config will specify the expected input tensor shape. It may be `(batch, channels, time_steps, height, width)` rather than `(batch, channels, height, width)`. Check the model card carefully.

2. **Band ordering:** Prithvi expects HLS bands in a specific order (Blue, Green, Red, Narrow NIR, SWIR 1, SWIR 2). The Sen1Floods11 dataset uses Sentinel-1 SAR (different modality). The fine-tuned checkpoint handles this mapping -- do NOT manually remap bands.

3. **TerraTorch dependency:** The fine-tuned checkpoints may require `terratorch` for loading. If `terratorch` doesn't install cleanly on Jetson (ARM64), you may need to extract the raw PyTorch state_dict and load it manually into the model architecture.

4. **TensorRT dynamic shapes:** If the model supports variable input sizes, use `--minShapes`, `--optShapes`, `--maxShapes` flags with `trtexec`. For fixed 512x512, this isn't needed.

5. **Memory limits:** Jetson Orin Nano has 8GB shared CPU/GPU memory. If the model + data doesn't fit, try:
   - Reduce batch size to 1
   - Use FP16 inference even in PyTorch (`model.half()`)
   - Process tiles sequentially rather than batched

6. **Warm-up:** Always discard the first 5-10 inference calls. The first call includes CUDA kernel compilation and is not representative.

7. **Reproducibility:** Set `torch.manual_seed(42)` and `torch.backends.cudnn.deterministic = True` for reproducible results.

---

## PR / Commit Convention

- Branch: `task-01-flood` and `task-02-burn-scar`
- One PR per task with results
- Commit messages: `task-01: [description]` or `task-02: [description]`
- Include results table in PR description

---

## Questions / Blockers

If you hit any of these, document the issue and move on to what you can:
- [ ] TerraTorch won't install on Jetson ARM64
- [ ] Checkpoint input shape doesn't match dataset format
- [ ] TensorRT conversion fails (specific layer not supported)
- [ ] INT8 calibration requires training data access
- [ ] GPU OOM on 512x512 tiles

Document the error message, what you tried, and move to the next step. Don't get stuck.

---

## Timeline

| Day | Activity |
|-----|----------|
| Day 1 | Jetson setup: JetPack, Python env, verify CUDA/TensorRT |
| Day 2 | Task 1: Download checkpoint + dataset, run FP32 baseline, collect accuracy metrics |
| Day 3 | Task 1: ONNX export, TensorRT FP16 build, run optimized inference, power measurement |
| Day 4 | Task 2: Download checkpoint + dataset, run full pipeline (FP32 + ONNX + TRT FP16) |
| Day 5 | Both tasks: Generate visual outputs, compile results tables, create PR |
